var/spack/repos/builtin/packages/curl/package.py

```diff
@@ -146,31 +146,53 @@ class Curl(NMakePackage, AutotoolsPackage):
 
     @classmethod
     def determine_version(cls, exe):
-        curl = Executable(exe)
-        output = curl("--version", output=str, error=str)
-        match = re.match(r"curl ([\d.]+)", output)
+        curl_config = Executable(exe)
+        output = curl_config("--version", output=str, error=str)
+        match = re.match(r"libcurl ([\d.]+)", output)
         return match.group(1) if match else None
 
     @classmethod
     def determine_variants(cls, exes, version):
         for exe in exes:
             variants = ""
             curl = Executable(exe)
-            output = curl("--version", output=str, error=str)
-            if "nghttp2" in output:
-                variants += "+nghttp2"
-            protocols_match = re.search(r"Protocols: (.*)\n", output)
-            if protocols_match:
-                protocols = protocols_match.group(1).strip().split(" ")
-                if "ldap" in protocols:
-                    variants += "+ldap"
-            features_match = re.search(r"Features: (.*)\n", output)
-            if features_match:
-                features = features_match.group(1).strip().split(" ")
-                if "GSS-API" in features:
-                    variants += "+gssapi"
-            # TODO: Determine TLS backend if needed.
-            # TODO: Determine more variants.
+            output = curl("--features", "--protocols", "--configure", "--ssl-backends", output=str)
+            # Some recipes require these variants. Flag them for use or rejection by concretisation
+            for feature in ("nghttp2", "libssh2"):
+                variants += f"+{feature}" if f"with-{feature}" in output else f"~{feature}"
+
+            variants += "+libidn2" if "IDN" in output else "~libidn2"
+            # No recipe depends on these variants, but kept for now:
+            variants += "+gssapi" if "GSS-API" in output else "~gssapi"
+            variants += "+ldap" if "LDAP" in output else "~ldap"
+
+            # Julia requests a curl with tls=mbedtls
+            tls = (
+                "openssl"
+                if "OpenSSL" in output
+                else (
+                    "mbedtls"
+                    if "mbedTLS" in output
+                    else (
+                        "gnutls"
+                        if "GnuTLS" in output
+                        else (
+                            "nss"
+                            if "NSS" in output
+                            else (
+                                "sspi"
+                                if is_windows
+                                else (
+                                    "secure_transport"
+                                    if sys.platform == "darwin"
+                                    else "detection-failed"  # Fail the detection of external curl
```

Comment:
This is an invalid value for the variant. If we fail to detect the `tls` value, can we return `None` and does that then result in a failed detection?

Reply:
Both, None and any invalid variant let the detection fail.
The point is to not make it fail with "None" the log from `spack -d`, which is hard to trace back to understand when looking at log.
A string that tells what going on was the idea.

And yes, the idea is to fail the detection when `tls` could be identified because it's better to not use a libcurl that uses a wrong TLS backend than to use it and break the build.